### ISSUE TITLE
--html-details cannot locate sources

### ISSUE THREAD

**Issue by DexterMagnific** [gcovr/gcovr#368]
Hi,

I have a problem with the --html-details options. It seems like gcovr tries to search for sources inside the build dir.

Here is my dir structure:
```
root/
  src/
    lib/
    test/
    build/debug/{lib|test}
```

the build dir is inside the src/ dir

gcda and gcno files are located inside src/build/debug/{lib|test}

with --html-details options, gcovr seems to search for source files inside src/build/debug/{lib|test}.

I have the following exception:

```
Traceback (most recent call last):
  File "C:\Python37\Scripts\gcovr-script.py", line 11, in <module>
    load_entry_point('gcovr==4.2', 'console_scripts', 'gcovr')()
  File "c:\python37\lib\site-packages\gcovr\__main__.py", line 254, in main
    print_reports(covdata, options, logger)
  File "c:\python37\lib\site-packages\gcovr\__main__.py", line 373, in print_reports
    generator(covdata, output.value, options)
  File "c:\python37\lib\site-packages\gcovr\html_generator.py", line 249, in print_html_report
    errors='replace') as INPUT:
FileNotFoundError: [Errno 2] No such file or directory: 'src\\build\\lib\\debug\\myfile.cpp'
```

the command line run from root/:
'''
gcovr.exe --html --html-details -r . -o coverage.html --gcov-executable /c/Qt/Tools/mingw730_32/bin/gcov.exe -v
'''

Can you please help with this ?

Thanks

**Comment by steven-ro** [comment]
The problem occurs for me as well, if I have my source directory at /src with other paths e.g. /workingdir or /home/username/src it works

**Comment by latk** [comment]
I fully believe that something is broken here, because gcovr's path handling depends a lot on guesses and heuristics. And those guesses often get it wrong. So thank you for adding these useful data points :)

What might help here is running gcovr from your build dir, especially if you are using cmake:

``` bash
cd src/build/debug; gcovr -r ../.. --html-details -o ../../../coverage.html ...
```

Do you at least get the correct files in other report formats, e.g. the default text report?

**Comment by DexterMagnific** [comment]
Running gcovr from the build dir did not fix the problem.

As for the text report, I do get a summary (as for the HTML report), but as you can't browse the source file, I do not know if it is relevant.

Maybe I did not specify that I do get an HTML report with per-file coverage statistics. The only problem is that I cannot browse the files because the paths are not correct.

**Comment by latk** [comment]
I think I have a hunch. Assuming that you are using gcovr 4.2, could you try editing the file `gcovr/html_generator.py` around line 248:

``` diff
         data['ROWS'] = []
-        currdir = os.getcwd()
-        os.chdir(options.root_dir)
-        with io.open(data['FILENAME'], 'r', encoding=options.source_encoding,
+        with io.open(f, 'r', encoding=options.source_encoding,
                      errors='replace') as INPUT:
             for ctr, line in enumerate(INPUT, 1):
                 data['ROWS'].append(
                     source_row(ctr, line.rstrip(), cdata.lines.get(ctr))
                 )
-        os.chdir(currdir)
```

Background: the `data['FILENAME']` variable is set to `options.root_filter.sub('', f)`, which just chops off the root path from the front. This should usually be fine, but there's no good reason to do this for a filename that we have to open. In particular, there's no strong guarantee that the `root_filter` only matches at the start of the filename – this needs to be checked in more detail. This part of gcovr still carries a bit of technical debt :)

Even if this isn't the exact problem, I think the root cause has to be nearby.

It would be helpful to have a small test case that reproduces the problem (hello world style source code plus a script of makefile to run gcc & gcovr), but if this is too Windows-specific that might be difficult.

**Comment by DexterMagnific** [comment]
Unfortunately no change, no matter if gcovr is run from the root dir or the build dir.

Before the patch:
```
FileNotFoundError: [Errno 2] No such file or directory: 'build\\lib\\debug\\MyFile.cpp'
```

After the patch:
```
FileNotFoundError: [Errno 2] No such file or directory: 'C:\\Users\\XXX\MyProject\\src\\build\\lib\\debug\\MyFile.cpp'
```

So the file is still searched in the build dir. the difference is only on the stripped 'root' prefix

**Comment by latk** [comment]
Thank you for helping investigate this, that was really helpful in ruling out some problem areas, although I still don't know what the actual problem is. If you can reproduce the issue in a small sample project this would make it possible to do deeper debugging, but I won't promise any timeline.

If you need a workaround: use gcovr to output JSON data, write a script to fix the path names in the JSON, and then use gcovr to turn the JSON into the HTML report. Alternatively, please consider other coverage tools such as lcov.

**Comment by k10blogger** [comment]
I use a two step process for a workaround for this issue. The first step uses the 'gcov' to map the data to the c file and then at root i use the gcovr to generate HTML from that data. This is possible because during my build process i isolate the files and create a batch file for the same. This happens automatically for me.

What i fail to understand is how is gcovr guessing the path and from where?
Cant we have a separate path list for sources and objects, so that we can append the sources path to one and append the object path to other. 
The gcovr just needs to take each object in object path and look for the source in source path. Instead of it guessing the path?

**Comment by latk** [comment]
Hi @k10blogger! Running the gcov tool manually tends to work very well, except when multiple object files contain coverage data for the same source file, e.g. inline functions in header files.

Gcovr needs to figure out from which directory the gcov tool should be executed. Generally, this should be exactly the directory from which the compiler (gcc) was invoked. I'm not entirely sure why this is necessary, but it seems that the gcno files contain relative pathnames to the source files. Gcovr uses a brute-force approach and just tries out lots of directories until gcov successfully produces output, see the `process_datafile()` function. https://github.com/gcovr/gcovr/blob/1f17d4427c301f2a2a8bad50991f161a0e45aa5c/gcovr/gcov.py#L514

This issue indicates that although the gcov tool produces correct output, gcovr misunderstands the source file name as reported by gcov. This is likely an error at the gcov–gcovr interface, which won't be fixed by giving a source path to gcovr. GCC 9 added working-directory metadata to the gcov reports which *might* fix the issue, but gcovr doesn't make use of it yet.

**Comment by thoni56** [comment]
I have something similar. In my case I have proper sources in the root directory and unittests in a subdirectory. I'm sending objects to a subdirectory `.objects` in both directories.

In the textual output I get references to source files from the unittests but with the `.objects` directory inserted, like:

    unittests/.objs/yylex_tests.c

Those do not exists, so that's probably why the HTML output fails.

I'm running gcovr 4.2 on WSL  4.4.0-18362-Microsoft.

**Comment by Spacetown** [comment]
@thoni56 You are using two different root directories for compiling but only one for running gcovr. The files produced by gcov contains the path used when compiling and the guessing of the path use the wrong root.
https://github.com/gcovr/gcovr/blob/39f5a82fb742300ea09adffa8f70b96a17f79061/gcovr/gcov.py#L186-L211
If you don't change to the sub directory for compiling `unit` but give the files with the sub directory everything is fine.

**Comment by thoni56** [comment]
@Spacetown Fair enough, is there any other way to merge the two data sets? I really don't want to bring up all my unittest compilation into the top level Makefile...

**Comment by Spacetown** [comment]
You can create a JSON for each executable using the Sam working directory as the compiler. This JSONs can be merged to  one HTML report.

**Comment by thoni56** [comment]
Say what?!?

I'm guessing that by "using the Sam working directory as the compiler" you mean that I should create two separate Json reports in the two directories where the compiles are run, right? Then those JSON files can be merged into a single report? Ok, I'lll try that.

**Comment by Spacetown** [comment]
Sorry for my bad English but you got it.

**Comment by f18m** [comment]
I'm hitting the same problem of original reporter of this bug... since indeed I have a similar tree structure for my project...

**Comment by f18m** [comment]
@Spacetown any help we can provide on fixing this problem? What's the status? is there a work-in-progress MR perhaps? Any hint on where to start the debugging? I would like to help if I can... Thanks a lot!

**Comment by Spacetown** [comment]
@f18m Can someone check if the generation of two separate JSON files and merging thme together will work?

**Comment by f18m** [comment]
> @f18m Can someone check if the generation of two separate JSON files and merging thme together will work?

Hi @Spacetown I tried something like that: instead of running 

`
gcovr -r . --verbose --print-summary --html-details $OUTDIR/coverage.html
`

I tried using:

`
gcovr -r .  --json >$OUTDIR/coverage.json
gcovr --add-tracefile $OUTDIR/coverage.json --html-details $OUTDIR/coverage.html
`

but it fails as well - I think the coverage.json file contains wrong source file paths and thus the html output generator will keep looking for the wrong source files also in this case.

**Comment by Spacetown** [comment]
I've created a test, there you can see the commands and from which directory the have to be called.
All gcovr calls have to use the same working directory. The second one has a different root directory (last argument). This is needed because the paths in the JSON are relative to the working directory of gcovr.
If this is the solution for your problem I'll create a PR for this test.

**Comment by f18m** [comment]
Hi @Spacetown ,

> I've created a test, there you can see the commands and from which directory the have to be called.
> All gcovr calls have to use the same working directory. The second one has a different root directory (last argument). This is needed because the paths in the JSON are relative to the working directory of gcovr.

Sorry it's not clear to me what you mean. You're saying that I should invoke gcovr in every single subfolder of my project that contains source code? that's going to be quite cumbersome...  also I think that --html-details should work out of the box in a single run, without the need to combine together multiple --json output files, right?

Also gcov >= 9.1 has a "current_working_directory" JSON entry in gcov JSON output (https://gcc.gnu.org/onlinedocs/gcc/Invoking-Gcov.html#Invoking-Gcov) that could be used to avoid such issue, right?

>If this is the solution for your problem I'll create a PR for this test.

But just to understand: in your feature branch you did not change gcovr internal path handling but just added a test to show that gcovr works if it's invoked multiple times from different folders, right?

Thanks!!

**Comment by Spacetown** [comment]
> Sorry it's not clear to me what you mean. You're saying that I should invoke gcovr in every single subfolder of my project that contains source code? that's going to be quite cumbersome... also I think that --html-details should work out of the box in a single run, without the need to combine together multiple --json output files, right?

Not in each source directory. You have to call gcovr for each executable from a common base directory giving the root directory used for compiling the executable.
Reason:
The JSON file contains paths relative to the working directory of gcovr and the gcov output contains paths relativ to the compiler working directory.

> Also gcov >= 9.1 has a "current_working_directory" JSON entry in gcov JSON output (https://gcc.gnu.org/onlinedocs/gcc/Invoking-Gcov.html#Invoking-Gcov) that could be used to avoid such issue, right?

We use the text output and support also old versions of gcc. At work we use 4.5 and 4.8.

> But just to understand: in your feature branch you did not change gcovr internal path handling but just added a test to show that gcovr works if it's invoked multiple times from different folders, right?

Yes

**Comment by ribtoks** [comment]
I'm also hitting the same issue. `--html` works, but `--html-details` fails with `FileNotFoundError: [Errno 2] No such file or directory`.

Any tips how to overcome this?

**Comment by Spacetown** [comment]
@ribtoks Run gcovr in the same directory than GCC. The `--html-details` fails because with this option the source files are read and added to for the report.

**Comment by ribtoks** [comment]
@Spacetown Thanks for the suggestion, but it does not work.

I have a following codebase structure:

```
src/
  - corelib/
    - models/
      - myclass.h
      - myclass.cpp
```

When I'm running `cd src && gcovr -r ./ `, the report looks like this:

```
Directory: ./
------------------------------------------------------------------------------
File                                       Lines    Exec  Cover   Missing
------------------------------------------------------------------------------
src/corelib/mymodel.cpp
```

Note: missing `/models/` component.

After I try: `gcovr -r ./ --html-details -o coverage/` and I get the error (surprise-surprise!) `FileNotFoundError: [Errno 2] No such file or directory: 'corelib/mymodel.cpp'`

Of course it's not found, it's missing `/models/` subdirectory. And there's no way I can fix it, so it's a bug!

**Comment by Spacetown** [comment]
@ribtoks Do you realy call the compiler in the `src` directory? You can also use the option `--debug` to get more informations about the path from the `gcov` output and the searched directories.

**Comment by ribtoks** [comment]
@Spacetown Yes, I _really_ call the compiler in the `src/` as in `cd src && qmake && make`. There's no `--debug` option in `gcovr`, but there's `--verbose`. It shows the same incorrect behavior:

```
Finding source file corresponding to a gcov data file
  currdir      /builds/projectname/plugins/model-release-plugin
  gcov_fname   /tmp/tmp8esj7xiw/Models#propertymodel.cpp.gcov
               ['        -', '    0', 'Source', 'Models/propertymodel.cpp\n']
  source_fname /builds/projectname/plugins/model-release-plugin/src/model-release-corelib/propertymodel.gcda
  root         /builds/projectname/plugins/model-release-plugin
  fname        /builds/projectname/plugins/model-release-plugin/src/model-release-corelib/propertymodel.cpp
Parsing coverage data for file /builds/projectname/plugins/model-release-plugin/src/model-release-corelib/propertymodel.cpp
```

Note that `propertymodel.cpp` is in subdirectory `Models`, but the result `fname` for gcovr does not have this subdirectory.

**Comment by Spacetown** [comment]
Sorry, `--verbose` was the correct one. Do you run the compiler with `qmake` and `gcovr` with `make`?
For me it seems that `qmake` runs the compiler not in `src` but in the sub directory `corelib`. Can you check this?

**Comment by ribtoks** [comment]
@Spacetown , `qmake` only generates the `Makefile`, but yes, `gcc` runs inside `corelib` directory. Thanks! Do I now have to launch gcovr from every subdirectory separately to generate a separate reports? Is there a normal way to launch it only once?

**Comment by Spacetown** [comment]
@ribtoks Yes, you have to run gcovr for each directory in the src directory and generate a trace file. This trace files can be merged together with a separate call.

Or can you configure qmake to keep the working directory?

**Comment by ribtoks** [comment]
@Spacetown No, I doubt I can tweak `qmake` behavior. Can you please provide some pointers to documentation on how to generate and merge trace files?

**Comment by Spacetown** [comment]
@ribtoks Here you are: https://gcovr.com/en/stable/guide.html#combining-tracefiles

**Comment by ribtoks** [comment]
For anybody looking for solution in future: I did not come up with anything convenient. Manually searching for correct directories (with `find`), calling `gcovr` inside them and then merging relevant coverage files ~~is nuts~~ leads to a lot of boilerplate and copy-paste across projects. Hopefully this will be supported in the software itself.

**Comment by Spacetown** [comment]
@ribtoks Do you have a solution how to guess the correct working directory of the `gcc` used by `qmake` or a self written make? The only idea I have is to throw the ball back to the user and add a option to replace the code of the guessing function https://github.com/gcovr/gcovr/blob/c0f1da2f9f1fa78bbe4dc093da0c953d2d3a3550/gcovr/gcov.py#L196

**Comment by ribtoks** [comment]
@Spacetown no, this I don't know. But maybe you can use a file search from root directory and hope that there will be no similarly named files in the same file tree.

**Comment by Spacetown** [comment]
I've prepared a quick test branch (https://github.com/Spacetown/gcovr/tree/recursive_search_file), please can you check if this is working?

**Comment by ribtoks** [comment]
@Spacetown My gcovr setup runs in CI/CD. I will try your branch, but it will take me a bit to replicate the setup locally. Thanks a lot for your help and prompt changes! I will be back sometime soon.

**Comment by blackliner** [comment]
I just added `CCACHE_BASEDIR=/home/user/.conan/data/package/....some_hashed_dir/` and ended up with all of my sourcefiles not being found anymore 🤔

Is `gcov` or `gcovr` relying on information stored in the built files by the compiler or something similar?

**Comment by latk** [comment]
@blackliner  Yes, GCC creates .gcno files along the .o object files. Information in the .gcno is used to associate coverage data with source code locations. Changing the location of these files will break gcov/gcovr, though it might be possible to work around some of the issues with gcov's “cross-profiling” feature (aka GCOV_PREFIX). See also <https://github.com/gcovr/gcovr/issues/481#issuecomment-808914953>.

**Comment by blackliner** [comment]
But I did not change any location, just `ccache` makes some paths that start with `CCACHE_BASEDIR` from absolute to relative.

**Comment by blackliner** [comment]
Seems my issue relates to https://github.com/ccache/ccache/issues/219

**Comment by ocroquette** [comment]
We just hit this bug too. The weird thing is that it happened by updating CMake 3.20 to 3.22. We use CMake and ninja to instrument the compiler to generate coverage information and then call gcov. With CMake 3.20, --html-details works fine. With CMake 3.22, we get the same error, because gcov is looking for cpp files in the build tree instead of the source tree. The --root option is correctly set but this doesn't help. It is still a mistery why the bug is triggered.
As a (bad) workaround, we removed --html-details for now.

**Comment by Spacetown** [comment]
@ocroquette Which version of gcovr are you using?

**Comment by ocroquette** [comment]
Hi @Spacetown I can reproduce the issue with 4.2 and 5.0

**Comment by Spacetown** [comment]
In [cmake 3.21](https://cmake.org/cmake/help/latest/release/3.21.html#id22) there is one change mentioned which maybe change the behaviour of `gcc`:
```
The Ninja Generators now pass source files and include directories to the compiler using absolute paths. This makes diagnostic messages and debug symbols more consistent, and matches the Makefile Generators.
```

If you use latest master branch, can you upload the build log where `gcovr` with the additional option `--verbose`? Then we can see the call of the compiler, the gcovr call and the search of the files?

**Comment by Spacetown** [comment]
Canc you try the LATEST master branch because we changed with #525 the heuristic to find the correct working director. See https://github.com/gcovr/gcovr/pull/525#issuecomment-964577837 for details.

**Comment by ocroquette** [comment]
Quick update...

- Using the master branch of gcovr does not help
- Updating Ninja to 1.10.2 does not help

Related CMake changes:

- https://gitlab.kitware.com/cmake/cmake/-/issues/13894
- https://gitlab.kitware.com/cmake/cmake/-/merge_requests/6148

I cannot send a log file for now sorry, I need to create a minimal example without internal information first.

**Comment by Spacetown** [comment]
Is it possible to strip the log to the gcov file which reference the file which isn't found? Maybe you can also send the file via mail if this is possible for you. You can also try the option `-fprofile-abs-path`of `gcc` version 9 or newer.
In our docker tests we use cmake version 3.22.1 and there is no problem.

**Comment by ocroquette** [comment]
@Spacetown I sent the log files to your GMX address. -fprofile-abs-path does not help,  gcovr still fails with the same error.

**Comment by ocroquette** [comment]
@Spacetown found the issue. We were setting  "--source-prefix" without a good reason, which, combined with the CMake update, led to the issue. Everything is fine without "--source-prefix".

Now I wonder if this Github issue should still be open?

**Comment by Spacetown** [comment]
Can this be checked with the branch of #597?

**Comment by Spacetown** [comment]
No progress for a while. If the problem persist, please reopen.

**Comment by ddomnik** [comment]
Hello, I think I face the same or a very similar problem on Windows using Cygwin for my catch2 UnitTests.
Folder structure looks like this:
```
myComponent/
- myComponent.c
- myComponent.h
- unit_test
   - myComponent.test.cpp
   - CMakeLists
   - build
```
I run `cmake ..` and then `make` from inside the build dir. For gcovr, it does not seem to help from which directory I run the command, it will always fail.

It searches for the source files in this directory: `CMakeFiles/myComponent.dir/cygdrive/c/Users/USERNAME/Documents/esp/project/myComponent/myComponent.c`
this ofc will not work, as in this folder are only the compiled `.c.o`, `.c.o.d`, `.c.gcda`, `.c.dcno` files.

How to tell gcovr which dir it should use?
